# 📗 Day 4 — Selecting Data

**Class record.** This is the notebook we worked through together, cleaned up and annotated.
Every section now has a short explanation under it, so you can re-read this without the lecture.

**Covered today:** selecting columns · `iloc` · `loc` · boolean indexing · `loc` with a condition ·
`value_counts()` · changing values with `loc` · the Titanic hands-on · missing values

**Not covered today:** the Bonus section at the end (`isin`, `between`, `sort_values`,
`reset_index`, `set_index`). It is marked clearly. Read it if you like, but nothing depends on it.

📂 **Datasets:** `tips.csv` and `Titanic-Dataset.csv` — keep both in the same folder as this notebook.

### How to use this notebook

- Run it top to bottom. Cells depend on the ones above them.
- Blank cells with only a `#` comment are **yours to fill**. That is the practice.
- ⭐ marks the parts people get wrong most often. Slow down there.

In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv('tips.csv')          # keep the file next to this notebook
# df = pd.read_csv('/Users/shailesh/Desktop/vedam_Sem3_ml/Data/tip.csv')

df.head()

In [ ]:
df.head()

---
## 1️⃣ Selecting Columns

> **Flow:** `df['col']` → Series · `df[['col']]` → DataFrame · many columns · new column

In [ ]:
# selecting many columns
x=df['day']
print(type(x))
# print(x.shape)
y=df[['day']]
print(type(y))
# print(y.shape)
y=df[['total_bill','day','tip']]
print(y.head())

In [ ]:
# making a new column
#Add a column called `bill_per_person` — the bill divided by the number of people
df['country']="India"
df['bill_per_person']=round(df['total_bill']/df['size'],2)

df.head()

**What just happened**

- `df['day']` gives a **Series** — one column, one dimension. `type()` proves it.
- `df[['day']]` gives a **DataFrame** — a table. The extra brackets are a *list of column names*.
- `df[['total_bill','day','tip']]` — same idea, three names in the list.
- `df['country'] = "India"` — assigning a single value fills the whole column with it.
- `df['bill_per_person'] = df['total_bill'] / df['size']` — assigning a calculation runs it
  row by row automatically. No loop anywhere.

⭐ One bracket = column. Two brackets = table. This difference comes back all day.

### 🧩 Try these

**1.1** Add a column called `total_paid` — the bill plus the tip. Write the full line.

```python
df['total_paid'] = ____
```

In [ ]:
# 1.1  — total_paid = bill + tip
df['total_paid']=df['tip']+df['total_bill']
df.head()

**1.2** Run both of these and print `type()` of each.
One is a single column, one is a table that happens to have one column. Which is which?
```python
df['tip']
df[['tip']]
```

In [ ]:
# 1.2

### 🔧 Break my code

Both lines are wrong. Find the mistake without running them.


```python
df.shape()
df['total_bill', 'tip']
```

💬 Type the mistake in chat — but press Enter only when I say **go**.

In [ ]:
# what is wrong?

In [24]:
df.head(6)

,total_bill,tip,sex,smoker,day,time,size,country,bill_per_person,total_paid
0,16.99,1.01,Female,No,Sun,Dinner,2,India,8.49,18.00
1,10.34,1.66,Male,No,Sun,Dinner,3,India,3.45,12.00
2,21.01,3.50,Male,No,Sun,Dinner,3,India,7.00,24.51
3,23.68,3.31,Male,No,Sun,Dinner,2,India,11.84,26.99
4,24.59,3.61,Female,No,Sun,Dinner,4,India,6.15,28.20
5,25.29,4.71,Male,No,Sun,Dinner,4,India,6.32,30.00


---
## 2️⃣ `iloc` — selecting by position

You did this in NumPy. Here we use it to pick **rows**.

> **Flow:** one row → many rows → rows and columns together

In [27]:
df.head(3)

,total_bill,tip,sex,smoker,day,time,size,country,bill_per_person,total_paid
0,16.99,1.01,Female,No,Sun,Dinner,2,India,8.49,18.00
1,10.34,1.66,Male,No,Sun,Dinner,3,India,3.45,12.00
2,21.01,3.50,Male,No,Sun,Dinner,3,India,7.00,24.51


In [ ]:
# iloc — one row, then a slice of rows
# df.iloc[4] for single row
# df.iloc[6:10:2]
# 6,7,8,10
# df.iloc[[6,7,8,10]] # fancy indexing 
df.iloc[(df['bill_per_person']>4).values]

# df.iloc[2,[1,3,5]]

NotImplementedError: iLocation based boolean indexing on an integer type is not available

**What just happened**

`iloc` means *integer location* — it only understands **numbers**.

```python
df.iloc[4]           # one row, by position
df.iloc[6:10:2]      # a slice, with a step
df.iloc[[6,7,8,10]]  # fancy indexing — a list of positions
df.iloc[2, [1,3,5]]  # row 2, columns at positions 1, 3 and 5
```

The last line in the demo was a different trick:

```python
df.iloc[(df['bill_per_person'] > 4).values]
```

A condition normally goes inside plain `df[...]`. To hand it to `iloc` you have to strip
the labels off first with `.values`, leaving a bare True/False array. You will almost never
need this — plain `df[condition]` is the normal way — but it shows what `iloc` really wants:
positions and nothing else.

In [ ]:
# iloc — rows and columns by position


### 🧩 Try these

**2.1** Fill the blanks.


```python
df.iloc[____]      # the 6th row
df.iloc[____]      # the last row
df.iloc[____]      # the first 5 rows
```

In [ ]:
# 2.1  — the 6th row, the last row, the first 5 rows

**2.2** Show only `day` and `total_bill`. Fill the blank.
```python
print(list(df.columns))     # run this first

df.iloc[10:15, ____]
```

In [30]:
df.head()

,total_bill,tip,sex,smoker,day,time,size,country,bill_per_person,total_paid
0,16.99,1.01,Female,No,Sun,Dinner,2,India,8.49,18.00
1,10.34,1.66,Male,No,Sun,Dinner,3,India,3.45,12.00
2,21.01,3.50,Male,No,Sun,Dinner,3,India,7.00,24.51
3,23.68,3.31,Male,No,Sun,Dinner,2,India,11.84,26.99
4,24.59,3.61,Female,No,Sun,Dinner,4,India,6.15,28.20


In [31]:
# 2.2

df.iloc[10:15,[4,0]]

,day,total_bill
10,Sun,10.27
11,Sun,35.26
12,Sun,15.42
13,Sun,18.43
14,Sun,14.83


---
## 3️⃣ `loc` — selecting by label

> **Flow:** row by label → rows and columns by name → **the slicing difference** ⭐

In [37]:
# loc — one row by its label
df.loc[5:10,'total_bill':'time']

,total_bill,tip,sex,smoker,day,time
5,25.29,4.71,Male,No,Sun,Dinner
6,8.77,2.00,Male,No,Sun,Dinner
7,26.88,3.12,Male,No,Sun,Dinner
8,15.04,1.96,Male,No,Sun,Dinner
9,14.78,3.23,Male,No,Sun,Dinner
10,10.27,1.71,Male,No,Sun,Dinner


**What just happened**

```python
df.loc[5:10, 'total_bill':'time']
```

Two things at once here:

1. `5:10` are **labels**, not positions — and `loc` **includes** the end. That is **6** rows, not 5.
2. `'total_bill':'time'` slices the **columns by name**, giving you every column from
   `total_bill` through `time` in the order they sit in the table.

⭐ This is the thing `iloc` cannot do. `iloc` would need `df.iloc[5:11, 0:6]` — and you would
have to count the columns yourself.

In [ ]:
# loc — rows and columns by name


### ✋ Guess first, then run

`df.iloc[2:5]` and `df.loc[2:5]`



In [38]:
print(df.iloc[2:5])
print(df.loc[2:5])


   total_bill   tip     sex smoker  day    time  size country  \
2       21.01  3.50    Male     No  Sun  Dinner     3   India   
3       23.68  3.31    Male     No  Sun  Dinner     2   India   
4       24.59  3.61  Female     No  Sun  Dinner     4   India   

   bill_per_person  total_paid  
2             7.00       24.51  
3            11.84       26.99  
4             6.15       28.20  
   total_bill   tip     sex smoker  day    time  size country  \
2       21.01  3.50    Male     No  Sun  Dinner     3   India   
3       23.68  3.31    Male     No  Sun  Dinner     2   India   
4       24.59  3.61  Female     No  Sun  Dinner     4   India   
5       25.29  4.71    Male     No  Sun  Dinner     4   India   

   bill_per_person  total_paid  
2             7.00       24.51  
3            11.84       26.99  
4             6.15       28.20  
5             6.32       30.00  


**The answer: different.**

- `df.iloc[2:5]` → **3** rows (2, 3, 4). It stops *before* 5.
- `df.loc[2:5]` → **4** rows (2, 3, 4, 5). It *includes* 5.

⭐ `loc` includes the end because labels are names, and it makes no sense to ask for
"everything up to but not including the row called 5". `iloc` follows normal Python slicing.

Remember it as: **loc = label = includes. iloc = index = excludes.**

In [ ]:
# my guess:

# now run both


### 🧩 Try these

**3.1** Find the row labelled 100. Which day was it, and what was the tip?

In [39]:
# 3.1  — the row labelled 100: which day, and what tip?
df.loc[100,['day','tip']]

day    Fri
tip    2.5
Name: 100, dtype: object

**3.2** Both lines should give the same 11 rows. Fill the blank.

```python
df.loc[10:20]        # 11 rows
df.iloc[10:____]     # make this 11 too
```

In [ ]:
# 3.2  — make iloc give the same 11 rows as loc[10:20]


**3.3** Rows 5 to 8, but only `day` and `tip` — asked for by name, not by number.

In [ ]:
# 3.3

In [ ]:
df.describe()

---
## 4️⃣ Boolean Indexing

So far you knew **which** row you wanted. Now you don't — you only know **what must be true** about it.

> **Flow:** the True/False list → use it to filter → count → two conditions with `&` and `|` → flip it with `~`

In [40]:
df.head(2)

,total_bill,tip,sex,smoker,day,time,size,country,bill_per_person,total_paid
0,16.99,1.01,Female,No,Sun,Dinner,2,India,8.49,18.0
1,10.34,1.66,Male,No,Sun,Dinner,3,India,3.45,12.0


In [44]:
# the condition on its own — True/False for every row
df[(df['day']=='Sun') & (df['size']>3)].shape[0]

df[(df['day']=='Sun') | (df['time']=='Dinner')].shape[0]

176

**What just happened**

A condition on its own does not filter anything — it produces a **True/False value for every row**:

```python
df['day'] == 'Sun'      # 244 True/False values
```

Put that inside `df[...]` and Pandas keeps only the True ones. Put `.sum()` on it and you get
a count, because True counts as 1 and False as 0.

Two conditions:

```python
df[(df['day']=='Sun') & (df['size']>3)]     # &  = both must be true
df[(df['day']=='Sun') | (df['time']=='Dinner')]  # |  = either one
```

⭐ **Brackets around each condition are not optional.** `&` is evaluated before `>` in Python,
so `df['size'] > 2 & df['tip'] > 4` tries to do `2 & df['tip']` and dies.

In [ ]:
# use it to filter


In [ ]:
# count how many matched


In [ ]:
# two conditions — & and |


#### Flipping a condition with `~`

You can find the Sunday bills. Now find every bill that was **not** on a Sunday — without using `!=`.

`~` reads as **"not"**. Every True becomes False, every False becomes True.

In [45]:
# ~ flips the whole condition
df[~(df['day']=="Sun")].head()

,total_bill,tip,sex,smoker,day,time,size,country,bill_per_person,total_paid
19,20.65,3.35,Male,No,Sat,Dinner,3,India,6.88,24.00
20,17.92,4.08,Male,No,Sat,Dinner,2,India,8.96,22.00
21,20.29,2.75,Female,No,Sat,Dinner,2,India,10.14,23.04
22,15.77,2.23,Female,No,Sat,Dinner,2,India,7.88,18.00
23,39.42,7.58,Male,No,Sat,Dinner,4,India,9.86,47.00


`~` means **not**. It flips every True to False and every False to True.

`df[~(df['day']=='Sun')]` → **168** rows, because 76 of the 244 bills were on Sunday.

Same result as `df[df['day'] != 'Sun']`, but `~` works on any condition, however long —
you wrap the whole thing once instead of rewriting each comparison.

### 🧩 Try these

**4.1** Bills where the tip was above 6. Fill the blanks, then count them.

```python
df[df[____] ____ 6]
```

💬 Drop **just the number** in chat when you have it.

In [ ]:
# 4.1  — bills where the tip was above 6, then count them


**4.2** Every bill that was **not** on a Sunday. Use `~`, then count them.

In [ ]:
# 4.2

**4.3** Fill the blanks.
```python
df[(df['total_bill'] > 30) ____ (df['tip'] > 5)]      # both true
df[(df['time'] == 'Dinner') ____ (df['size'] > 4)]    # either one true
```

In [ ]:
# 4.3  — fill in & and |

### 🔧 Break my code

All three are wrong. Find the mistakes.
```python
df[df['tip' > 6]]
df[df['size'] > 2 & df['tip'] > 4]
df.iloc[0, 'tip']
```

💬 Chat on my **go**. One line number + what's wrong.

In [49]:
# what is wrong?
# df[df['tip']>6]
# df[(df['size'] > 2) & (df['tip'] > 4)]
df.iloc[0, 'tip']

ValueError: Location based indexing can only have [integer, integer slice (START point is INCLUDED, END point is EXCLUDED), listlike of integers, boolean array] types

**The three mistakes**

1. `df[df['tip' > 6]]` — the bracket closes in the wrong place, so Python tries to compare
   the *text* `'tip'` with the number 6. Correct: `df[df['tip'] > 6]`.
2. `df[df['size'] > 2 & df['tip'] > 4]` — missing brackets, as above.
   Correct: `df[(df['size'] > 2) & (df['tip'] > 4)]`.
3. `df.iloc[0, 'tip']` — `iloc` never accepts a name.
   Correct: `df.loc[0, 'tip']`, or `df.iloc[0, 1]` if you insist on numbers.

---
## 5️⃣ Putting it together — `loc` with a condition

**Rows chosen by a condition. Columns chosen by name. One line.**

This is the most common line in all of Pandas.

In [ ]:
# loc[ condition , columns ]


**5.1** Bills with a tip above 6 — show only `day` and `tip`. Fill both blanks.
```python
df.loc[____, ____]
```

In [ ]:
# 5.1  — tip above 6, showing only day and tip

**5.2** What was the average `total_bill` on Sundays? ⭐

In [ ]:
# 5.2  — average total_bill on Sundays


---
## 6️⃣ Counting groups — `value_counts()`

You have been filtering one day at a time. Now the owner walks in and asks:

*"How many bills did we get on each day?"*

You could write four separate conditions. Or you could write one line.

> **Flow:** count every value in a column → count again after filtering

In [51]:
df.head()

,total_bill,tip,sex,smoker,day,time,size,country,bill_per_person,total_paid
0,16.99,1.01,Female,No,Sun,Dinner,2,India,8.49,18.00
1,10.34,1.66,Male,No,Sun,Dinner,3,India,3.45,12.00
2,21.01,3.50,Male,No,Sun,Dinner,3,India,7.00,24.51
3,23.68,3.31,Male,No,Sun,Dinner,2,India,11.84,26.99
4,24.59,3.61,Female,No,Sun,Dinner,4,India,6.15,28.20


In [52]:
# how many bills on each day
df['day'].value_counts()


day
Sat     87
Sun     76
Thur    62
Fri     19
Name: count, dtype: int64

**What just happened**

`value_counts()` counts how many times each value appears in a column — in one line, sorted
biggest first. Without it you would write one condition per day and add them up by hand.

You can filter first and then count, which is where it gets genuinely useful:

```python
df[df['time'] == 'Dinner']['day'].value_counts()
```

⭐ Notice **Thur = 1**. The restaurant served exactly one dinner on a Thursday all week.
That is not a bug in your code.

### 🧩 Try these

**6.1** How many bills were Lunch, and how many were Dinner?

In [53]:
# 6.1
df['time'].value_counts()


time
Dinner    176
Lunch      68
Name: count, dtype: int64

**6.2** ⭐ Out of the **Dinner** bills only, how many were on each day?

Filter first, then count. One day's number is going to look like a mistake. It isn't.

In [58]:
# 6.2
df[df['time']=='Dinner']['day'].value_counts()

day
Sat     87
Sun     76
Fri     12
Thur     1
Name: count, dtype: int64

---
## 7️⃣ Changing values, not just reading them

Everything until now only **looked** at the data. Nothing changed.

Same `loc`, same condition — but now the line ends with an `=`.

> **Flow:** give the whole column a default → overwrite it only where the condition is true → check

The restaurant decides: **every table of 6 or more gets a free dessert.**

In [ ]:
# step 1 — new column free_dessert, everyone starts as 'No'


In [ ]:
# step 2 — change it to 'Yes' only where the condition is true, then check with value_counts()

### 🧩 Try these

**7.1** Make a column called `bill_type`. Everyone starts as `'normal'`.
Bills above 40 become `'big'`. How many big ones are there?

In [60]:
# 7.1
df['bill_type']='normal'
df.loc[df['total_bill']>40,'bill_type']='Big'
print(df.head())

   total_bill   tip     sex smoker  day    time  size country  \
0       16.99  1.01  Female     No  Sun  Dinner     2   India   
1       10.34  1.66    Male     No  Sun  Dinner     3   India   
2       21.01  3.50    Male     No  Sun  Dinner     3   India   
3       23.68  3.31    Male     No  Sun  Dinner     2   India   
4       24.59  3.61  Female     No  Sun  Dinner     4   India   

   bill_per_person  total_paid bill_type  
0             8.49       18.00    normal  
1             3.45       12.00    normal  
2             7.00       24.51    normal  
3            11.84       26.99    normal  
4             6.15       28.20    normal  


In [61]:
df["bill_type"].value_counts()

bill_type
normal    234
Big        10
Name: count, dtype: int64

**What just happened — the pattern to memorise**

```python
df['bill_type'] = 'normal'                       # 1. everyone gets a default
df.loc[df['total_bill'] > 40, 'bill_type'] = 'Big'   # 2. overwrite where the condition is true
df['bill_type'].value_counts()                   # 3. check
```

This is the same `loc[condition, column]` line you used for *reading* — the only difference
is the `=` at the end. Reading and writing use identical syntax.

⭐ Careful with spelling: the question said `'big'`, the code above writes `'Big'`. Pandas will
not warn you — they are simply two different labels. Pick one and stay consistent.

**7.2** ⭐ The waiter says row 5's tip was typed in wrong — it should be `5.00`.
Print that one value, fix only that one cell, print it again to confirm.

In [ ]:
# 7.2

---
# 🚢 Titanic

April 1912. The Titanic sank. Some people survived, some did not.
Who survived was not random. Let's find out who.

Each box has two questions. I will solve the first one. You solve the second one.

🔀 **Breakout rooms from here.** Pairs, 8 minutes per box. One of you shares the screen,
the other reads the question out loud. Come back with a number.

In [ ]:
t = pd.read_csv('Titanic-Dataset.csv')
# t = pd.read_csv('/Users/shailesh/Desktop/vedam_Sem3_ml/Data/Titanic-Dataset.csv')

In [ ]:
# warm-up — you just learnt value_counts(). How many passengers in each class?

In [ ]:
t.head(10)

,PassengerId,Survived,Pclass,Name,Sex,Age,SibSp,Parch,Ticket,Fare,Cabin,Embarked
0,1,0,3,"Braund, Mr. Owen Harris",male,22.0,1,0,A/5 21171,7.2500,NaN,S
1,2,1,1,"Cumings, Mrs. John Bradley (Florence Briggs Th...",female,38.0,1,0,PC 17599,71.2833,C85,C
2,3,1,3,"Heikkinen, Miss. Laina",female,26.0,0,0,STON/O2. 3101282,7.9250,NaN,S
3,4,1,1,"Futrelle, Mrs. Jacques Heath (Lily May Peel)",female,35.0,1,0,113803,53.1000,C123,S
4,5,0,3,"Allen, Mr. William Henry",male,35.0,0,0,373450,8.0500,NaN,S
5,6,0,3,"Moran, Mr. James",male,NaN,0,0,330877,8.4583,NaN,Q
6,7,0,1,"McCarthy, Mr. Timothy J",male,54.0,0,0,17463,51.8625,E46,S
7,8,0,3,"Palsson, Master. Gosta Leonard",male,2.0,3,1,349909,21.0750,NaN,S
8,9,1,3,"Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)",female,27.0,0,2,347742,11.1333,NaN,S
9,10,1,2,"Nasser, Mrs. Nicholas (Adele Achem)",female,14.0,1,0,237736,30.0708,NaN,C


In [64]:
t.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

**Before you start: look at the column names.**

This file uses **capital letters** — `Survived`, `Pclass`, `Sex`, `Age`, `SibSp`, `Parch`,
`Fare`, `Cabin`, `Embarked`. `t['age']` will fail; `t['Age']` works.

Two columns worth knowing:

- `SibSp` — siblings and spouse aboard
- `Parch` — parents and children aboard

There is no "travelling alone" column, so you build it yourself:
`(t['SibSp'] == 0) & (t['Parch'] == 0)`.


### How many passengers were in first class?



In [67]:
# How many passengers were in first class?
t[t['Pclass']==1].shape[0]

216

In [ ]:
#  How many passengers were above age 60?


###  How many men were travelling alone?



In [68]:
# How many men were travelling alone?
t[(t['Sex']=='male')&(t['SibSp']==0)& (t['Parch']==0)].shape[0]

411

In [69]:
# How many children under 10 were in third class?
t[(t['Age'] < 10) & (t['Pclass'] == 3)].shape[0]

42

### How many men over 40 survived?



In [71]:
# How many men over 40 survived?
t[(t['Sex'] == 'male') & (t['Age'] > 40) & (t['Survived'] == 1)].shape[0]

18

In [ ]:
# How many women under 30 in second class survived?


###  Out of all the men, what percentage survived?



In [ ]:
# Out of all the men, what percentage survived?


In [ ]:
# Out of everyone who paid more than 50 for their ticket, what percentage survived?

###  Out of all the **women**, what percentage survived?

Put this number next to the men's number and read them out loud together.

In [ ]:
# Out of all the women, what percentage survived?

---
### ⚠️ Stop. Something was wrong with all of that.

Every single answer above used the `Age` column.

**Out of 891 passengers, how many have no age written down at all?**

In [ ]:
t['Age'].isnull().sum()

In [74]:
t.shape

(891, 12)

In [ ]:
# how many passengers were older than 40? and how many were 40 or younger?
# print both


print(t[t['Age']>40].shape[0])
print(t[t['Age']<=40].shape[0])


150
564


In [75]:
t['Age'].mean()

np.float64(29.69911764705882)

`29.70`. But divided by **what**?

Pandas skipped every blank and divided by 714, not 891. It made a decision on your behalf —
*ignore the missing ones* — and it never told you.

In [ ]:
# add those two numbers together. It should come to 891. Does it?

**150 + 564 = 714.**

There were **891** people on that ship.

177 of them just … weren't there. Not counted as over 40. Not counted as under 40.
Not counted at all.

And look at your screen. No error. No red text. No warning.
Your answer came out clean, and confident, and **wrong**.

In [ ]:
# one more — what was the average age on board?

Pandas quietly divided by 714, not by 891.

It already made a decision on your behalf — *skip the blanks* — and it never asked you.

In [ ]:
# and it is not only age. Check every column at once.

`isnull()` turns the whole table into True/False — True wherever a value is missing.
`.sum()` then counts the Trues per column.

Result: `Age` **177**, `Cabin` **687**, `Embarked` **2**, everything else clean.

### 🔚 So — tomorrow.

And it is not only `Age`. Look at `Cabin`: **687 blanks out of 891.**
More than three quarters of that column simply does not exist.

You know how to *filter* now. That part is finished.

Tomorrow's question is a different one: **what do you do with a hole?**

You get three options, and every one of them changes your answer:

- **throw those rows away** — and lose 177 real people
- **fill the blanks with something** — but with *what*, and who gave you the right to decide?
- **leave them alone** — and stay quietly wrong, exactly like you were five minutes ago

There is no correct option here. There is only a choice you can **defend** when someone asks you why.

**That's Day 5 — Data Cleaning. See you tomorrow.** 🧹

---
# ⚠️ Bonus — NOT COVERED IN CLASS

**We did not reach this section today.** Nothing in the assignment uses it and nothing
tomorrow depends on it. Read it only if you are curious.

Everything below does something you can **already do**. These are just shorter ways.
A shortcut only makes sense once you know the long way.

## `isin()` — instead of a long chain of `|`

The long way:
```python
df[(df['day'] == 'Sat') | (df['day'] == 'Sun')]
```

In [ ]:
# the short way — isin()


**B1** Find all Lunch bills from Thursday or Friday, using `isin()`.

In [ ]:
# B1  — Lunch bills from Thursday or Friday, using isin()

## `between()` — instead of two conditions with `&`

The long way:
```python
df[(df['total_bill'] > 20) & (df['total_bill'] < 30)]
```

In [ ]:
# the short way — between()


**B2** Find bills where the tip was between 3 and 5.

In [ ]:
# B2  — tips between 3 and 5, using between()

## `sort_values()` and `nlargest()`

In [ ]:
# sort by a column, largest first


**B3** Find the 5 biggest tips.

**B4** Sort by `total_bill`, largest first, and save it as `top`.
Now run `top.iloc[0]` and `top.loc[0]`. **Different rows?** Explain why in a comment. ⭐

In [ ]:
# B3  — the 5 biggest tips

In [ ]:
# B4  — sort by total_bill (largest first), save as top
#      then compare top.iloc[0] and top.loc[0] — why are they different?

## `reset_index()` — fixing labels after filtering

In [ ]:
sat = df[df['day'] == 'Sat']

sat.iloc[0]      # works — first Saturday bill
# sat.loc[0]     # uncomment this. It fails. Why?

In [ ]:
# fix it with reset_index(drop=True), then try .loc[0] again

**B5** Filter the Thursday bills into `thur`. Now make `thur.loc[0]` work.

In [ ]:
# B5

## `set_index()` — the moment `loc` finally makes sense

`loc` picks by **label**. All day your labels were 0, 1, 2, 3… boring numbers,
which made `loc` look like a slower `iloc`.

Watch what happens when the labels are actual words.

⚠️ Save it as **`d2`**, not `df`. We don't want to disturb the original table.

In [ ]:
# make 'day' the label — save it as d2, NOT df

In [ ]:
# now pick every Sunday row by its name

**B6** Using `d2`, pull out all the Saturday rows. Does `d2.iloc[0]` still work?

In [ ]:
# B6

---
## 📌 Recap — everything from today on one screen

### Picking columns

| You write | You get |
|---|---|
| `df['tip']` | a **Series** — one column |
| `df[['tip']]` | a **DataFrame** — a table with one column |
| `df[['day','tip']]` | a table with two columns |
| `df['new'] = ...` | creates a new column |

### `loc` vs `iloc` ⭐

| | `loc` | `iloc` |
|---|---|---|
| Picks by | **label** (name) | **position** (number) |
| Slice end | **included** | **not included** |
| Columns given as | names | numbers |
| Takes a condition | **yes** | no |

`df.loc[10:20]` gives **11** rows. `df.iloc[10:20]` gives **10**. This is the single most
common mistake in Pandas.

`loc` can also slice **columns** by name: `df.loc[5:10, 'total_bill':'time']`.

### Boolean indexing

Write a condition, get True/False for every row, keep the True ones.

```python
df[df['tip'] > 6]              # filter
(df['tip'] > 6).sum()          # count — True counts as 1
df[(df['size'] > 2) & (df['tip'] > 4)]     # both true
df[(df['time'] == 'Dinner') | (df['size'] > 4)]   # either true
df[~(df['day'] == 'Sun')]      # NOT
```

⭐ Every condition needs **its own brackets**. `&` runs before `>`, so leaving them out breaks it.

### `loc` with a condition — the most used line in Pandas

```python
df.loc[df['tip'] > 6, ['day','tip']]        # rows by condition, columns by name
df.loc[df['day'] == 'Sun', 'total_bill'].mean()
```

### Counting

```python
df['day'].value_counts()                          # count each value
df[df['time'] == 'Dinner']['day'].value_counts()  # filter first, then count
```

### Changing values

```python
df['free_dessert'] = 'No'                          # default for everyone
df.loc[df['size'] >= 6, 'free_dessert'] = 'Yes'    # overwrite where true
df.loc[5, 'tip'] = 5.00                            # fix one single cell
```

### What broke everything at the end

`Age` has **177** blanks. 150 were over 40, 564 were 40 or under — that is **714**, not 891.
The missing 177 were silently dropped from every answer. No error appeared.
`Cabin` is worse: **687** blanks.

**🔜 Day 5 — Data Cleaning:** those 177 missing ages · `dropna` vs `fillna` · duplicates ·
sorting · and a dataset made from **your own** answers 📊

---
## 🧠 Exit quiz — answer in chat before you leave

No running code. From memory.

1. `df.loc[10:20]` gives how many rows? And `df.iloc[10:20]`?
2. Fix this line: `df[df['size'] > 4 & df['tip'] > 5]`
3. You saved the Saturday bills as `sat`. Now `sat.loc[0]` fails. Why?
4. One line: which day had the most bills?

### Answers to the exit quiz

1. `df.loc[10:20]` → **11** rows. `df.iloc[10:20]` → **10** rows.
2. Each condition needs brackets: `df[(df['size'] > 4) & (df['tip'] > 5)]`.
3. Filtering keeps the **original labels**. The first Saturday bill is labelled 19, not 0 —
   so there is no row labelled 0 to fetch. Use `.iloc[0]` instead.
4. `df['day'].value_counts().idxmax()` → `'Sat'`.

In [ ]:
# quiz answers

## https://shorturl.at/QB5Uj